### PR TITLE
[OpenAI] Update function_call sample + Prepare for the beta 6 release

### PR DESCRIPTION
**Pull request by mssfang**
- Update function_call sample
- Prepare for the new beta 6 release.

### CODE REVIEW

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionCallingSample.java [hunk 1]
```diff
@@ -46,62 +48,65 @@ public static void main(String[] args) {
...
+                .setDescription("The city and state, e.g. San Francisco, CA");
+
+        FunctionProperties format = new FunctionProperties()
+                .setType("string")
+                .setEnumString(Arrays.asList("celsius", "fahrenheit"))
+                .setDescription("The temperature unit to use. Infer this from the users location.");
+
+        Map<String, FunctionProperties> props = new HashMap<>();
+        props.put("location", location);
+        props.put("format", format);
```
Comment (jpalvarezl) [inline sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionCallingSample.java:84]: I think that you should rename `format` to `unit` as this is a bit confusing. This is for requesting the temperature in either Fahrenheit or Celsius, right?

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/models/FunctionParameters.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+     * @param requiredPropertyNames List of required properties.
+     * @return Object itself.
+     */
+    public FunctionParameters setRequiredPropertyNames(List<String> requiredPropertyNames) {
+        this.requiredPropertyNames = requiredPropertyNames;
+        return this;
+    }
+
+    @Override
+    public String toString() {
```
Comment (jpalvarezl) [inline sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/models/FunctionParameters.java:89]: Nit: I am not a big fan of overriding `toString`, not sure if this is a practice we should encourage in our samples. If you agree, I would rename this method to `prettyPrint` or something along those lines.

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionToolCallingSample.java [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+                                getNDayWeatherForecastToolFunction
+                        )
+                );
+
+        // We can force the model to use a specific function, for example get_n_day_weather_forecast by using the
+        // function_call argument. By doing so, we force the model to make assumptions about how to use it.
+        chatCompletionsOptions.setToolChoice(BinaryData.fromObject(getNDayWeatherForecastToolFunction));
+
+        // We can also force the model to not use a function at all. By doing so we prevent it from producing a proper function calling.
+        // chatCompletionsOptions.setToolChoice(BinaryData.fromString("none"));
```
Comment (jpalvarezl) [inline sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionToolCallingSample.java:78]: This comment is a bit confusing, as it doesn't explain the code around it, it rather comments on a different aspect of the feature. If you want, you could add a separate sample with `ToolChoice::None`, otherwise I would remove it.

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionToolCallingSample.java [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+                .setProperties(props);
+    }
+
+    private static List<ChatRequestMessage> handleFunctionToolResponse(List<ChatChoice> choices,
+                                                                      List<ChatRequestMessage> chatMessages) {
+        for (ChatChoice choice : choices) {
+            ChatResponseMessage choiceMessage = choice.getMessage();
+
+            List<ChatCompletionsToolCall> toolCalls = choiceMessage.getToolCalls();
+            if (toolCalls != null) {
```
Comment (jpalvarezl) [inline sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionToolCallingSample.java:168]: I would rather add a check here for `finish_reason.TOOL_CALL` as well (or instead)

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionToolCallingSample.java [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+        String deploymentOrModelId = "{azure-open-ai-deployment-model-id}";
+
+        // Function call is supported for model versions with suffix `-0613` and later.
+        OpenAIClient client = new OpenAIClientBuilder()
+                .endpoint(endpoint)
+                .credential(new AzureKeyCredential(azureOpenaiKey))
+                .buildClient();
+
+        singleFunctionCallingSample(client, deploymentOrModelId);
+        parallelFunctionCallingSample(client, deploymentOrModelId);
```
Comment (jpalvarezl) [inline sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionToolCallingSample.java:55]: I think this should be separated into 2 samples. It's really hard to follow what you are trying to focus on here. I would suggest splitting this into `ChatCompletionsSingleToolCall` and `ChatCompletionsParallelToolCall`. 

Here you seem to be conflating `function_call` with `tool_call`. Depending how you setup your request, these will produce different `finishe_reasons` , AFAIK.

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionCallingSample.java [hunk 6]
```diff
@@ -46,62 +48,65 @@ public static void main(String[] args) {
...
-        ChatCompletionsOptions chatCompletionOptions = new ChatCompletionsOptions(chatMessages)
-            .setFunctionCall(FunctionCallConfig.AUTO)
-            .setFunctions(functions);
-
-        ChatCompletions chatCompletions = client.getChatCompletions(deploymentOrModelId, chatCompletionOptions);
-        List<ChatRequestMessage> chatMessages2 = handleFunctionCallResponse(chatCompletions.getChoices(), chatMessages);
+        ChatCompletions chatCompletions = client.getChatCompletions(deploymentOrModelId,
+                new ChatCompletionsOptions(chatRequestMessages)
+                        .setFunctionCall(FunctionCallConfig.AUTO)
+                        .setFunctions(functions));
```
Comment (jpalvarezl) [inline sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionCallingSample.java:61]: This sample is about `FunctionCall` and it's correctly named. I think the ones below need to be updated to reflect the difference between a `tool_call` and a `function_call`.

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionCallingSample.java [hunk 7]
```diff
@@ -26,11 +28,11 @@
 import java.util.Map;
 
 /**
- * Sample demonstrates how to get chat completions using function call.
+ * Sample demonstrates how to get chat completions using function calling.
  */
-public class FunctionCallSample {
+public class FunctionCallingSample {
```
Comment (jpalvarezl) [inline sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionCallingSample.java:33]: nit: I think that `FunctionCall` was better :D as it points the user directly to the class and it's not necessarily grammatically incorrect. 

nit 2: I would go as far as maybe renaming to `ChatCompletionsFunctionCallSample` to really drive the point that this is a `ChatCompletions` feature.

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionToolCallingSample.java [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+                                getNDayWeatherForecastToolFunction
+                        )
+                );
+        // Newer models like gpt-4-1106-preview or gpt-3.5-turbo-1106 can call multiple functions in one turn.
+        ChatCompletions chatCompletions = client.getChatCompletions(deploymentOrModelId, chatCompletionsOptions);
+        chatRequestMessages = handleFunctionToolResponse(chatCompletions.getChoices(), chatRequestMessages);
+
+        ChatCompletions chatCompletionsAnswer = client.getChatCompletions(deploymentOrModelId,
+                new ChatCompletionsOptions(chatRequestMessages));
+        getContentFromMessageResponse(chatCompletionsAnswer);
```
Comment (jpalvarezl) [inline sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionToolCallingSample.java:107]: I think that where this sample falls a bit short is in that it should also showcase what to do with the response, once the LLM requests the `tool_call` here. 

Then you could really showcase that you are calling things in parallel by doing something like:

```java
Flux.zip(
  Mono.fromCallable(() -> getCurrentWeatherToolFunction()),
  Mono.fromCallable(() -> getNDayWeatherForecastToolFunction()),
  Pair::new
).flatMap( pair -> {
  String first = pair.getFirst();
  String second = pair.getSecond();
  // ... the rest of the code
  client.getChatCompletions(...)
}).subscribe(...);
```

Comment (jpalvarezl) [reply sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionToolCallingSample.java:107]: Here I am pulling `Pair<T, K>` out of thin air. Here is a copilot implementation of that:

```java
    class Pair<T, K> {
        private T first;
        private K second;

        public Pair(T first, K second) {
            this.first = first;
            this.second = second;
        }

        public T getFirst() {
            return first;
        }

        public K getSecond() {
            return second;
        }
    }
```

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/models/FunctionProperties.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+     */
+    public FunctionProperties setEnumString(List<String> enumString) {
+        this.enumString = enumString;
+        return this;
+    }
+
+    @Override
+    public String toString() {
+        return "FunctionEntry [type=" + type + ", description=" + description + "]";
+    }
```
Comment (jpalvarezl) [inline sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/models/FunctionProperties.java:88]: Nit: Same comment as above. Maybe not necessarily overriding this and calling this function somethings else?

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionToolCallingSample.java (hunk 8 above)
Comment (jpalvarezl) [reply sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionToolCallingSample.java:107]: I am just using the names of the function as you have set them, but it seems a bit weird to call those functions like that. Maybe we could use different functions for this particular example.

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionToolCallingSample.java (hunk 4 above)
Comment (jpalvarezl) [reply sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionToolCallingSample.java:168]: I've written sync/async samples for this in case you need to check that out. [Link](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/usage/GetChatCompletionsToolCallSample.java)

Comment (jpalvarezl) [reply sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionToolCallingSample.java:168]: Also streaming (it's not pretty)

Comment (mssfang) [reply sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionToolCallingSample.java:168]: I saw there is a possible that having "finish_reason is STOPPED" but still have toolCalls values. 

image

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/models/FunctionParameters.java (hunk 2 above)
Comment (mssfang) [reply sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/models/FunctionParameters.java:89]: removed

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/models/FunctionProperties.java (hunk 9 above)
Comment (mssfang) [reply sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/models/FunctionProperties.java:88]: removed

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionToolCallingSample.java (hunk 5 above)
Comment (mssfang) [reply sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionToolCallingSample.java:55]: I would delete the single case since it is easy to convert to single from multiple

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionCallingSample.java (hunk 1 above)
Comment (jpalvarezl) [reply sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionCallingSample.java:84]: The `json` in the article describes the field as:

```json
"format": {
    "type": "string",
    "enum": ["celsius", "fahrenheit"],
    "description": "The temperature unit to use. Infer this from the users location.",
},
```

I think that maybe it's a mistake in the naming, I think we could fix it. Specially, because in our other samples we have similar mehtods and we call it `unit` instead of `format`. I would personally prefer that we align with our own samples in this regard.

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionToolCallingSample.java (hunk 4 above)
Comment (jpalvarezl) [reply sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionToolCallingSample.java:168]: Are those the choices from the `choiceMessage`. I am not sure I can tell exactly from the screen shot 🤔 I wasn't aware we could have a `finish_reason` other than `tool_call` in this case.

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionCallingSample.java (hunk 1 above)
Comment (mssfang) [reply sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/FunctionCallingSample.java]: Updated

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/ChatCompletionsFunctionCall.java [hunk 10]
```diff
@@ -46,58 +49,61 @@ public static void main(String[] args) {
...
-        return functionDefinition;
+    private static FunctionParameters getCurrentWeatherFunctionParameters() {
+        FunctionProperties location = new FunctionProperties()
+                .setType("string")
+                .setDescription("The city and state, e.g. San Francisco, CA");
+
+        FunctionProperties unit = new FunctionProperties()
+                .setType("string")
+                .setEnumString(Arrays.asList("celsius", "fahrenheit"))
+                .setDescription("The temperature unit to use. Infer this from the users location.");
```
Comment (srnagar) [inline sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/ChatCompletionsFunctionCall.java:81]:
```suggestion
                .setDescription("The temperature unit to use. Infer this from the user's location.");
```

**Review by srnagar — APPROVED** [review]
